rllib/core/rl_trainer/rl_trainer.py

```diff
@@ -83,9 +100,11 @@ def __init__(
         self.distributed = distributed
         self.in_test = in_test
 
-        # These are the attributes that are set during build
-        self._params: List[ParamType] = None
-        self._optimizers: List[Optimizer] = None
+        # These are the attributes that are set during build for properly applying
+        # optimizers and adding or removing modules.
+        self._optim_to_param: Dict[Optimizer, List[ParamRef]] = {}
+        self._param_to_optim: Dict[ParamRef, Optimizer] = {}
+        self._params: Dict[ParamRef, ParamType] = {}
```

Comment:
I would create a named_tuple for (ParamType, Optimizer).
btw, why do you need to know the mapping from param to optimizer?
sometimes book keeping too much stuff is error prone. for example, since there shouldn't be too many params, keeping a flat list then for loop it is often an acceptable option.

Reply:
The reason is that tensorflow's optimizers don't keep track of their parameters at the time of construction. You need to call optim.apply(gradient, param) for the first time to register params to an optim object. It's a tf induced design requirement. O.W. Just having the optimizer objects would suffice in torch.